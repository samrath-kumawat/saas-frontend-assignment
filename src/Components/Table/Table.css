/* .row{
    border: 1px solid #dddddd;
    text-align: left;
    padding: 8px;
} */
.tableHeader {
  font-size: 14px;
  padding: 12px 15px;
  background-color: #009879;
}

.tableData {
        padding: 12px 15px;
        font-size: 14px;
}
.tableDataRow{
    /* margin-top: 10px; */
    /* padding-top: 10px; */
    height: 50px;
    text-align: left;
        /* border-bottom: 1px solid #dddddd; */
        border-radius: 10px;
        box-shadow: 0 0 10px rgba(0, 0, 0, 0.12);
}

.thFirstCellRadius{
  border-top-left-radius: 10px;
  border-bottom-left-radius: 10px;
}
.thLastCellRadius{
  border-top-right-radius: 10px;
  border-bottom-right-radius: 10px;
}
.tableHeaderRow {

  /* background-color: #009879; */
  color: white;
  /* margin-top: 10px; */
  /* padding-top: 10px; */
  height: 60px;
  text-align: left;
      /* border-bottom: 1px solid #dddddd; */
      /* border-radius: 10px; */
      /* box-shadow: 0 0 10px rgba(0, 0, 0, 0.15); */
}
/* .table {
        border-collapse: collapse;
        margin-top: 50px;
        margin: 25px 0;
        width: 80%;
        box-shadow: 0 0 20px rgba(0, 0, 0, 0.15);
  } */
  .tableBody{
    width: 80%;
    border-spacing: 0px 10px;
    border-collapse: separate;
  }
  .container{
    margin: 50px;
    display: flex;
    flex-direction: column;
    justify-content: space-around;
    align-items: center;
  }