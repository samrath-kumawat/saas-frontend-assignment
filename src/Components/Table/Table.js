import React, { useEffect, useState } from 'react'
import { ITEM_PER_PAGE, WEB_API_URL } from '../../Constants';
import Pagination from '../Pagination/Pagination';
import './Table.css';

const Table = () => {

const [projectsData, setProjectsData] = useState([]);
const [currentPage, setCurrentPage] = useState(1);

const fetchProjects = async (url)=>{
    const response = await fetch(url);
    const data = await response.json();
    setProjectsData(data);
}

useEffect(()=>{
    fetchProjects(WEB_API_URL);
}, []);

const handlePageChange = (number)=>{
    setCurrentPage(number)
}

const startIndex = ITEM_PER_PAGE * (currentPage-1);
const endIndex = ITEM_PER_PAGE * currentPage;

  return (
    
    <div className='container'>
      <table className='tableBody'>
        <thead>
            <tr className='tableHeaderRow'>
                <th className='tableHeader thFirstCellRadius'>S.No.</th>
                <th className='tableHeader'>Percentage Funded</th>
                <th className='tableHeader thLastCellRadius'>Amount Pledged</th>
            </tr>
            <tr></tr>
            <tr></tr>
            </thead>
            <tbody>
            {projectsData.slice( startIndex, endIndex).map((val, key) => {
                return (
                    <tr className='tableDataRow' key={key} data-testid={`key-${val['s.no']}`}>
                        <td className='tableData' >{val['s.no']}</td>
                        <td className='tableData'>{val['percentage.funded']}</td>
                        <td  className='tableData'>{val['amt.pledged']}</td>
                    </tr>
                )
            })}
            </tbody>
        </table>
        {projectsData?.length > ITEM_PER_PAGE &&
            <Pagination
            totalItems={projectsData?.length}
            itemsPerPage={ITEM_PER_PAGE}
            currentPage={currentPage}
            onPageChange={handlePageChange}
        />
        }
        
        
    </div>
  )
}
export default Table
