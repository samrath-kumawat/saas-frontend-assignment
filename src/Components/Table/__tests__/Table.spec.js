import { render, screen, waitFor } from '@testing-library/react';
import Table from '../Table';

describe('Table Component', () => {
    
    global.fetch = jest.fn();
    global.setProjectsData = jest.fn();
    const mockResponse = [{ 's.no': 1, 'percentage.funded': 180, 'amt.pledged': 1345 }];

test('match snapshots', async () => {
    fetch.mockResolvedValueOnce({
        json: jest.fn().mockResolvedValueOnce(mockResponse)
      });
    const { asFragment } = render(<Table />);
    expect(asFragment()).toMatchSnapshot();
});
test('render component and fetch projects', async () => {
    
    fetch.mockResolvedValueOnce({
        json: jest.fn().mockResolvedValueOnce(mockResponse)
      });

    render(<Table />);
    await waitFor(() => screen.findByTestId('key-1')); 
    expect(screen.getByTestId('key-1')).toHaveTextContent('1');
});
test('render component and fetch projects more then 5', async () => {

    const mockResponse1 = [
        { 's.no': 1, 'percentage.funded': 180, 'amt.pledged': 1345 },
        { 's.no': 2, 'percentage.funded': 180, 'amt.pledged': 1345 },
        { 's.no': 3, 'percentage.funded': 180, 'amt.pledged': 1345 },
        { 's.no': 4, 'percentage.funded': 180, 'amt.pledged': 1345 },
        { 's.no': 5, 'percentage.funded': 180, 'amt.pledged': 1345 },
        { 's.no': 6, 'percentage.funded': 180, 'amt.pledged': 1345 },
        { 's.no': 7, 'percentage.funded': 180, 'amt.pledged': 1345 }
    ]
    
    fetch.mockResolvedValueOnce({
        json: jest.fn().mockResolvedValueOnce(mockResponse1)
      });

    render(<Table />);
    await waitFor(() => screen.findByTestId('key-1')); 
    expect(screen.getByTestId('key-1')).toHaveTextContent('1');
});
});