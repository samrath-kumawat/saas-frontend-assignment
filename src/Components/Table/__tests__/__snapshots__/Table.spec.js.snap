// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Table Component match snapshots 1`] = `
<DocumentFragment>
  <div
    class="container"
  >
    <table
      class="tableBody"
    >
      <thead>
        <tr
          class="tableHeaderRow"
        >
          <th
            class="tableHeader thFirstCellRadius"
          >
            S.No.
          </th>
          <th
            class="tableHeader"
          >
            Percentage Funded
          </th>
          <th
            class="tableHeader thLastCellRadius"
          >
            Amount Pledged
          </th>
        </tr>
        <tr />
        <tr />
      </thead>
      <tbody />
    </table>
  </div>
</DocumentFragment>
`;
